chore: Add decision record concerning downward compatibiity (#1797)
eclipse-tractusx/tractusx-edc: docs/development/decision-records/2025-02-12-dsp-downward-compatibility/README.md

# Handling of downward compatibility for a DSP version

## Decision

> DSP assets, like the catalog, which are forwarded directly to a caller through the management api, have to be kept downward compatible as long as the DSP version through which the asset is retrieved is supported. There is no compatibility requirement between two different DSP versions concerning such assets.

Reviewer: > "have to be kept downward compatible as long as the DSP version through which the asset is retrieved is supported"
Can this statement be explained in different terms?

Reviewer 2: yes, we need to be more specific, as here we're not talking about supporting a DSP version, but supporting the outcome of a specific EDC version based on a DSP version. 

e.g. the EDC connector versions subsequent to 0.7.7 need to being able to operate with EDC connector at version 0.7.7 as well as other third party app that were built to operate with version 0.7.7.
Specifying that the connector version implemented by the 0.7.7 connector is a "special" DSP 0.8 ([descibed here](https://github.com/International-Data-Spaces-Association/ids-specification/tree/v0.8)), not 100% compliant with the specs

> DSP assets, like the catalog, which are forwarded directly to a caller through the management api, have to be kept downward compatible as long as the DSP version through which the asset is retrieved is supported. There is no compatibility requirement between two different DSP versions concerning such assets.

Reviewer 2: "DSP assets" don't exist.
Asset is a concept internal to the EDC domain, in DSP it could be reflected in the "Dataset" term (from DCAT spec)

Relevant assets are:
- Catalog returned by...
- ...

## Rationale

> DSP assets which are directly received by a caller of a management api endpoint is data that is transfered between two dataspace participants and creates a direct relationship between the calling application and the provider connector. As this is the case, the breaking change requirements of Catena-X has to be fulfilled. The requirement states that a consumer can continuously consume data from the provider without the need to be forced to update his system due to a change of the provider service stack. This requires that such assets must not change in a way, that existing properties vanish from the returned data.

Reviewer 2: ```suggestion
DSP datasets which are obtained by a consumer third party app through the management api endpoint create a direct relationship between the calling application and the provider connector. As this is the case, the breaking change requirements of Catena-X has to be fulfilled. The requirement states that a consumer can continuously consume data from the provider without the need to be forced to update his system due to a change of the provider service stack. This requires that such assets must not change in a way, that existing properties vanish from the returned data.
```

this need to be rephrased:
> As this is the case
which case?

plus, a link to the "breaking change requirements of Catena-X" will be necessary

> As a consumer participant is capable to ensure the usage of a connector with a certain set of DSP versions supported, he is capable to use applications that are capable to support all potential data formats received by the different supported DSP versions. Therefore, the asset data models supported in two different DSP versions are not related and therefore no compatibility requirement is needed between those.

Reviewer 2: ```suggestion
As a consumer participant is capable to ensure the usage of a connector with a certain set of DSP versions supported, it is capable to use applications that are capable to support all potential data formats received by the different supported DSP versions. Therefore, the asset data models supported in two different DSP versions are not related and therefore no compatibility requirement is needed between those.
```

looks a bit convoluted, could this be simplified? Like "changes between different protocol versions are expected"

## Approach

> - The compatibility tests are enriched by tests that ensure the detection of changes in the corresponding DSP assets.

Reviewer 2: please explain the approach further

# Handling of downward compatibility for a DSP version

## Decision

For a given DSP protocol version *N*, there is no restriction concerning backward compatibility towards the previous

> versions (*N-1*) implementation of the DSP protocol.

Author: Ok, my point was, that this is a clear statement towards the Tractus-X community, that they cannot do any assumption at that point and have to explicitely deal with the version information. This might be implied by DSP, but by explicitely stating that in the DR decision part from us Tractus-X EDC provider the implication from upstream cannot be misunderstood.

> versions (*N-1*) implementation of the DSP protocol.

Author: I have adapted the rationale below. For me it is a part of the decision, you are right the DSP protocol works this way, but we could alternatively decide that we handle this with an layer in between which tries to cope with such breaking changes coming from the protocol implementation. As this makes no sense due to the reasons given in the rationale, we deliberately decide, that we forward the missing compatibility from upstream, as this can be handled with version awareness. This is our decision, inherited from upstream, but still our decision.

During the support of a certain DSP version *N*, the implementation has to be strictly backward compatible, i.e.,
for two successive connector versions *X-1* and *X* that support both the DSP version *N*, if one participant changes
from connector version *X-1* to version *X*, another dataspace participant using any compatible connector version and
third party tools that use this connector can continue to transfer data to and from that participant without the need

> to do any changes in his setup.

Author: But that is actually the requirement we phase currently. There are data structures that are forwarded from another EDC to third-party apps and the issue is, that they can expect backward compatibility. That is, why I explicitely excluded additions to the data model as breaking change.

That two EDC versions are interoperable concerning a defined DSP version is one thing, but it must not happen, that someone operates his app with his connector and the other day, he cannot use it because the data he receives from another connector has changed in a way, that expected fields are missing.

If he fails because some fields have been added, we can consider that as a bug in the app implementation, but not if the data structure is changed in a way, that expected fields are gone. That is the whole reason for this decision record.

> to do any changes in his setup.

Reviewer 2: Stating that there won't be any breaking change (eventually only additions) in a protocol version is more than enough, including that we will make third party app working as well will put us in a weird situation like the one happened in #1783 where the change was not breaking but that third party app had issue. 
We cannot guarantee that, in any case.

> to do any changes in his setup.

Author: I hope this new formulation makes the decision clearer

# Handling of downward compatibility for a DSP version

## Decision

For a given DSP protocol version *N*, there is no restriction concerning backward compatibility towards the previous
versions (*N-1*) implementation of the DSP protocol.

During the support of a certain DSP version *N*, the implementation has to be strictly backward compatible, i.e.,
for two successive connector versions *X-1* and *X* that support both the DSP version *N*, if one participant changes

> from connector version *X-1* to version *X*, any interaction with the new connector version *X* has no breaking
> behavior changes. Additions to exchanged data models are in general not seen as breaking changes.

Reviewer: ```suggestion
from connector version *X-1* to version *X*, any interaction with connector version *X-1* is kept intact and has no breaking
behavior. Additions to exchanged data models are in general not seen as breaking changes.
```

## Rationale

In general, the breaking change requirement of Catena-X has to be fulfilled as expressed in the
[Operating model](https://catenax-ev.github.io/docs/operating-model/how-life-cycle-management). The requirement states
that a dataspace participant can continuously exchange data with another particpant without the need to be forced to
update his system due to a change of the other participants service stack. This requires that data sent between
consumer and provider must not change in a breaking fashion.

A backward compatibility contraint between DSP versions is neither needed nor possible over time, as

> 1. a dataspace participant is able to control the connector version to use and thereby the DSP version offered in the
>    dataspace independently from other participants. The participant can choose a connector compatible to his
>    environment, so there is no need for a backward compatibility constraint.

Reviewer: ```suggestion
1. a dataspace participant should be able to control the connector version to use in his stack independently
 from other participants, as long as that version implements the DSP version(s) required by the dataspace 
 at that point. If all required DSP version(s) are supported, there is no need for a backward compatibility constraint. 
 Communication is still possible through previous DSP version(s).
```
